Add support for DS2482-800 and DS248x refactor (Cleaned up & Fixed)

tomwellnitz opened the pull request:
## Description

This PR adds full support for the **DS2482-800** (8-channel 1-Wire Master) and refactors the \`ds248x\` component to support multiple channels. It also includes support for **DS2482-100** and **DS2484**.

This is a cleaned-up and bug-fixed version of #11804.

### Changes & Improvements over #11804:
- **Cleaned up:** Removed all unrelated changes (climate, pid, cover, etc.) that were present in the original PR.
- **Bug Fix (Critical):** Fixed a logic error in \`select_channel\` that caused channel switching to fail on DS2482-800.
- **Bug Fix:** Fixed \`reset_hub\` to correctly sync the software channel state with the hardware state.
- **Safety:** Added a check to prevent channel switching commands on single-channel devices (DS2482-100, DS2484) to avoid conflicts with port configuration registers.
- **Code Quality:** Extracted into a clean branch based on current \`dev\`.

### Supported Hardware:
- DS2482-100 (Single Channel)
- DS2482-800 (8 Channels)
- DS2484 (Single Channel)

### Documentation:
- esphome/esphome-docs#5811

### Credits:
Original implementation by @vielleicht in #11804.
Refactoring and fixes by @tomwellnitz.

tomwellnitz committed:
Fix DS2482-800 channel selection and support single-channel devices

tomwellnitz committed:
Add CODEOWNERS for new components (including original author)

tomwellnitz committed:
Address review comments: Fix typos, modernize C++, fix copy-paste errors

lboue commented on esphome/components/ds248x/__init__.py:
```diff
@@ -0,0 +1,96 @@
+CONF_ACTIVE_PULLUP = "active_pullup"
+CONF_STRONG_PULLUP = "strong_pullup"
+CONF_CHANNEL_COUNT = "channel_count"
+CONF_CHANNEL = "channel"
```
```suggestion
```

lboue commented on esphome/const.py:
```diff
@@ -342,6 +342,7 @@ class PlatformFramework(Enum):
 CONF_DUMMY_RECEIVER_ID = "dummy_receiver_id"
 CONF_DUMP = "dump"
 CONF_DURATION = "duration"
```
```suggestion
CONF_DUMP = "dump"
CONF_DUMMY_RECEIVER = "dummy_receiver"
CONF_DUMMY_RECEIVER_ID = "dummy_receiver_id"
CONF_DURATION = "duration"
```

tomwellnitz committed:
Cleanup file permissions and update constants for DS248x and DS2438 components

tomwellnitz committed:
Add compilation tests and apply code fixes for ds248x

tomwellnitz committed:
Fix clang-tidy style violations in OneWire components

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

tomwellnitz committed:
Add Overdrive Speed and DS2484 support to ds248x component

tomwellnitz committed:
Fix: Initialize sleep_pin_ with nullptr in header

tomwellnitz committed:
Remove unrelated components ds18x and ds2438 from PR

tomwellnitz committed:
Fix: Remove reverted components from CODEOWNERS and tests

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

tomwellnitz committed:
Fix: Regenerate CODEOWNERS without ds18x and ds2438

tomwellnitz committed:
[ds248x] Address review comments: fix logic, add delays, init members

tomwellnitz committed:
[ds248x] Remove IRAM_ATTR from read_scratch_pad

tomwellnitz committed:
[ds248x] Add support for variable resolution, auto-SPU, and DS2484 timing

tomwellnitz committed:
Fix DS248x review comments: unsafe optionals, init vars, typos, tests

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

tomwellnitz committed:
Merge branch 'dev' into ds248x-cleanup

tomwellnitz committed:
Merge branch 'dev' into ds248x-cleanup

tomwellnitz committed:
Merge branch 'dev' into ds248x-cleanup

tomwellnitz committed:
Merge branch 'dev' into ds248x-cleanup

tomwellnitz committed:
Merge branch 'dev' into ds248x-cleanup

tomwellnitz committed:
Fix clang-tidy warnings and optimize DS248x driver

- Rename protected methods to have trailing underscores (clang-tidy)
- Remove unused sleep configuration methods
- Add parasitic mode support to sensor config
- Optimize loop structures

tomwellnitz committed:
Merge branch 'ds248x-cleanup' of myfork/ds248x-cleanup into ds248x-cleanup

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

tomwellnitz committed:
Merge branch 'dev' into ds248x-cleanup

tomwellnitz committed:
Fix whitespace in __init__.py (flake8)

tomwellnitz committed:
Merge branch 'ds248x-cleanup' of https://github.com/tomwellnitz/esphome into ds248x-cleanup

tomwellnitz committed:
Fix lint error: Add NOLINT to delay(100) in setup

tomwellnitz committed:
[ds248x] Fix SPU logic and implement 1-Wire Search

- Fix Strong Pullup (SPU) not being disabled after conversion.
- Implement 1-Wire Search algorithm to discover devices.
- Scan all channels on startup for DS2482-800.
- Fix crc8 namespace usage.

tomwellnitz committed:
Optimize: Dynamic conversion delay & support for DS18S20/DS1825/DS28EA00

tomwellnitz commented:
➡️ **New PR:** https://github.com/esphome/esphome/pull/12717

This PR is now obsolete and will be closed to keep the history clean.
